File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/OperatorBlockSyntax.cs
using System;
using System.ComponentModel;
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class OperatorBlockSyntax : MethodBlockBaseSyntax
	{
		internal OperatorStatementSyntax _operatorStatement;

		internal EndBlockStatementSyntax _endOperatorStatement;

		public OperatorStatementSyntax OperatorStatement => GetRedAtZero(ref _operatorStatement);

		public new SyntaxList<StatementSyntax> Statements
		{
			get
			{
				SyntaxNode red = GetRed(ref _statements, 1);
				return new SyntaxList<StatementSyntax>(red);
			}
		}

		public EndBlockStatementSyntax EndOperatorStatement => GetRed(ref _endOperatorStatement, 2);

		public override MethodBaseSyntax BlockStatement => OperatorStatement;

		public override EndBlockStatementSyntax EndBlockStatement => EndOperatorStatement;

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete.", true)]
		public new OperatorStatementSyntax Begin => OperatorStatement;

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete.", true)]
		public new EndBlockStatementSyntax End => EndOperatorStatement;

		internal OperatorBlockSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal OperatorBlockSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, OperatorStatementSyntax operatorStatement, SyntaxNode statements, EndBlockStatementSyntax endOperatorStatement)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OperatorBlockSyntax(kind, errors, annotations, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OperatorStatementSyntax)operatorStatement.Green, statements?.Green, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax)endOperatorStatement.Green), null, 0)
		{
		}

		public OperatorBlockSyntax WithOperatorStatement(OperatorStatementSyntax operatorStatement)
		{
			return Update(operatorStatement, Statements, EndOperatorStatement);
		}

		internal override SyntaxList<StatementSyntax> GetStatementsCore()
		{
			return Statements;
		}

		internal override MethodBlockBaseSyntax WithStatementsCore(SyntaxList<StatementSyntax> statements)
		{
			return WithStatements(statements);
		}

		public new OperatorBlockSyntax WithStatements(SyntaxList<StatementSyntax> statements)
		{
			return Update(OperatorStatement, statements, EndOperatorStatement);
		}

		public new OperatorBlockSyntax AddStatements(params StatementSyntax[] items)
		{
			return WithStatements(Statements.AddRange(items));
		}

		internal override MethodBlockBaseSyntax AddStatementsCore(params StatementSyntax[] items)
		{
			return AddStatements(items);
		}

		public OperatorBlockSyntax WithEndOperatorStatement(EndBlockStatementSyntax endOperatorStatement)
		{
			return Update(OperatorStatement, Statements, endOperatorStatement);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				0 => _operatorStatement, 
				1 => _statements, 
				2 => _endOperatorStatement, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				0 => OperatorStatement, 
				1 => GetRed(ref _statements, 1), 
				2 => EndOperatorStatement, 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitOperatorBlock(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitOperatorBlock(this);
		}

		public OperatorBlockSyntax Update(OperatorStatementSyntax operatorStatement, SyntaxList<StatementSyntax> statements, EndBlockStatementSyntax endOperatorStatement)
		{
			if (operatorStatement != OperatorStatement || statements != Statements || endOperatorStatement != EndOperatorStatement)
			{
				OperatorBlockSyntax operatorBlockSyntax = SyntaxFactory.OperatorBlock(operatorStatement, statements, endOperatorStatement);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(operatorBlockSyntax, annotations);
				}
				return operatorBlockSyntax;
			}
			return this;
		}

		public override MethodBlockBaseSyntax WithBlockStatement(MethodBaseSyntax blockStatement)
		{
			return WithOperatorStatement((OperatorStatementSyntax)blockStatement);
		}

		public override MethodBlockBaseSyntax WithEndBlockStatement(EndBlockStatementSyntax endBlockStatement)
		{
			return WithEndOperatorStatement(endBlockStatement);
		}

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete.", true)]
		public OperatorBlockSyntax WithBegin(OperatorStatementSyntax begin)
		{
			return WithOperatorStatement(begin);
		}

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete.", true)]
		public new OperatorBlockSyntax WithEnd(EndBlockStatementSyntax end)
		{
			return WithEndOperatorStatement(end);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/OperatorBlockSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System
Imports System.ComponentModel

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class OperatorBlockSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.MethodBlockBaseSyntax
		Friend _operatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax

		Friend _endOperatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete.", True)>
		Public Shadows ReadOnly Property Begin As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax
			Get
				Return Me.OperatorStatement
			End Get
		End Property

		Public Overrides ReadOnly Property BlockStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.MethodBaseSyntax
			Get
				Return Me.OperatorStatement
			End Get
		End Property

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete.", True)>
		Public Shadows ReadOnly Property [End] As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax
			Get
				Return Me.EndOperatorStatement
			End Get
		End Property

		Public Overrides ReadOnly Property EndBlockStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax
			Get
				Return Me.EndOperatorStatement
			End Get
		End Property

		Public ReadOnly Property EndOperatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax)(Me._endOperatorStatement, 2)
			End Get
		End Property

		Public ReadOnly Property OperatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax
			Get
				Return MyBase.GetRedAtZero(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax)(Me._operatorStatement)
			End Get
		End Property

		Public Shadows ReadOnly Property Statements As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)
			Get
				Return New SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)(MyBase.GetRed(Me._statements, 1))
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As Microsoft.CodeAnalysis.VisualBasic.SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal operatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax, ByVal statements As SyntaxNode, ByVal endOperatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OperatorBlockSyntax(kind, errors, annotations, DirectCast(operatorStatement.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.OperatorStatementSyntax), If(statements IsNot Nothing, statements.Green, Nothing), DirectCast(endOperatorStatement.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax)), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitOperatorBlock(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitOperatorBlock(Me)
		End Sub

		Public Shadows Function AddStatements(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			Return Me.WithStatements(Me.Statements.AddRange(items))
		End Function

		Friend Overrides Function AddStatementsCore(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.MethodBlockBaseSyntax
			Return Me.AddStatements(items)
		End Function

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Select Case i
				Case 0
					syntaxNode = Me._operatorStatement
					Exit Select
				Case 1
					syntaxNode = Me._statements
					Exit Select
				Case 2
					syntaxNode = Me._endOperatorStatement
					Exit Select
				Case Else
					syntaxNode = Nothing
					Exit Select
			End Select
			Return syntaxNode
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim operatorStatement As SyntaxNode
			Select Case i
				Case 0
					operatorStatement = Me.OperatorStatement
					Exit Select
				Case 1
					operatorStatement = MyBase.GetRed(Me._statements, 1)
					Exit Select
				Case 2
					operatorStatement = Me.EndOperatorStatement
					Exit Select
				Case Else
					operatorStatement = Nothing
					Exit Select
			End Select
			Return operatorStatement
		End Function

		Friend Overrides Function GetStatementsCore() As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)
			Return Me.Statements
		End Function

		Public Function Update(ByVal operatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax, ByVal statements As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax), ByVal endOperatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			Dim operatorBlockSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			If (operatorStatement <> Me.OperatorStatement OrElse statements <> Me.Statements OrElse endOperatorStatement <> Me.EndOperatorStatement) Then
				Dim operatorBlockSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.OperatorBlock(operatorStatement, statements, endOperatorStatement)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				operatorBlockSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, operatorBlockSyntax1, operatorBlockSyntax1.WithAnnotations(annotations))
			Else
				operatorBlockSyntax = Me
			End If
			Return operatorBlockSyntax
		End Function

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete.", True)>
		Public Function WithBegin(ByVal begin As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			Return Me.WithOperatorStatement(begin)
		End Function

		Public Overrides Function WithBlockStatement(ByVal blockStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.MethodBaseSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.MethodBlockBaseSyntax
			Return Me.WithOperatorStatement(DirectCast(blockStatement, Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax))
		End Function

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete.", True)>
		Public Shadows Function WithEnd(ByVal [end] As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			Return Me.WithEndOperatorStatement([end])
		End Function

		Public Overrides Function WithEndBlockStatement(ByVal endBlockStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.MethodBlockBaseSyntax
			Return Me.WithEndOperatorStatement(endBlockStatement)
		End Function

		Public Function WithEndOperatorStatement(ByVal endOperatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			Return Me.Update(Me.OperatorStatement, Me.Statements, endOperatorStatement)
		End Function

		Public Function WithOperatorStatement(ByVal operatorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			Return Me.Update(operatorStatement, Me.Statements, Me.EndOperatorStatement)
		End Function

		Public Shadows Function WithStatements(ByVal statements As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.OperatorBlockSyntax
			Return Me.Update(Me.OperatorStatement, statements, Me.EndOperatorStatement)
		End Function

		Friend Overrides Function WithStatementsCore(ByVal statements As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.MethodBlockBaseSyntax
			Return Me.WithStatements(statements)
		End Function
	End Class
End Namespace